File: backend/app/routers/exam_sessions.py
from typing import Optional

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_current_user, get_db
from app.models import User
from app.schemas.exam_session import (
    CreateExamSessionRequest,
    ExamErrorReportResponse,
    ExamSessionHistoryResponse,
    ExamSessionQuestionDetail,
    ExamSessionResponse,
    ExamSessionResultResponse,
    SubmitSessionAnswerRequest,
    SubmitSessionAnswerResponse,
)
from app.services import exam_session_service

router = APIRouter(prefix="/exam-sessions", tags=["exam-sessions"])


@router.post("", response_model=ExamSessionResponse, status_code=status.HTTP_201_CREATED)
async def create_session(
    data: CreateExamSessionRequest,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await exam_session_service.create_session(db, data.exam_id, user.id)


@router.get("/active", response_model=Optional[ExamSessionResponse])
async def get_active_session(
    exam_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await exam_session_service.get_active_session(db, exam_id, user.id)


@router.get("/history", response_model=ExamSessionHistoryResponse)
async def get_session_history(
    exam_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await exam_session_service.get_session_history(db, exam_id, user.id)


@router.get("/error-report", response_model=ExamErrorReportResponse)
async def get_error_report(
    exam_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await exam_session_service.get_error_report(db, exam_id, user.id)


@router.get("/{session_id}", response_model=ExamSessionResponse)
async def get_session(
    session_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await exam_session_service.get_session(db, session_id, user.id)


@router.get("/{session_id}/result", response_model=ExamSessionResultResponse)
async def get_session_result(
    session_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await exam_session_service.get_session_result(db, session_id, user.id)


@router.get("/{session_id}/questions/{order_index}", response_model=ExamSessionQuestionDetail)
async def get_session_question(
    session_id: int,
    order_index: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await exam_session_service.get_session_question(db, session_id, user.id, order_index)


@router.post(
    "/{session_id}/questions/{order_index}/submit",
    response_model=SubmitSessionAnswerResponse,
)
async def submit_session_answer(
    session_id: int,
    order_index: int,
    data: SubmitSessionAnswerRequest,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await exam_session_service.submit_session_answer(
        db, session_id, user.id, order_index, data.selected_option_ids
    )


@router.post("/{session_id}/complete", response_model=ExamSessionResultResponse)
async def complete_session(
    session_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await exam_session_service.complete_session(db, session_id, user.id)


@router.post("/{session_id}/abandon", status_code=status.HTTP_204_NO_CONTENT)
async def abandon_session(
    session_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    await exam_session_service.abandon_session(db, session_id, user.id)

File: workers-api/src/routes/exam-sessions.ts
import { Hono } from "hono";
import type { AppEnv } from "../types/env";
import { AppError } from "../lib/errors";
import {
  createSession,
  getSession,
  getActiveSession,
  getSessionQuestion,
  submitSessionAnswer,
  completeSession,
  abandonSession,
  getSessionResult,
  getSessionHistory,
  getErrorReport,
} from "../services/exam-session-service";

const sessionRoutes = new Hono<AppEnv>();

function requireExamIdQuery(c: { req: { query: (k: string) => string | undefined } }): number {
  const raw = c.req.query("exam_id");
  const id = Number(raw);
  if (!raw || isNaN(id)) throw new AppError(400, "exam_id query parameter is required");
  return id;
}

// POST /exam-sessions
sessionRoutes.post("/", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const body = await c.req.json<{ exam_id: number }>();
  if (!body.exam_id || typeof body.exam_id !== "number") {
    throw new AppError(400, "exam_id is required");
  }

  const result = await createSession(db, user.id, body.exam_id);
  return c.json(result, 201);
});

// GET /exam-sessions/active?exam_id=X
sessionRoutes.get("/active", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const examId = requireExamIdQuery(c);

  const result = await getActiveSession(db, user.id, examId);
  return c.json(result);
});

// GET /exam-sessions/history?exam_id=X
sessionRoutes.get("/history", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const examId = requireExamIdQuery(c);

  const result = await getSessionHistory(db, user.id, examId);
  return c.json(result);
});

// GET /exam-sessions/error-report?exam_id=X
sessionRoutes.get("/error-report", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const examId = requireExamIdQuery(c);

  const result = await getErrorReport(db, user.id, examId);
  return c.json(result);
});

// GET /exam-sessions/:id
sessionRoutes.get("/:id", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const sessionId = Number(c.req.param("id"));

  const result = await getSession(db, sessionId, user.id);
  return c.json(result);
});

// GET /exam-sessions/:id/result
sessionRoutes.get("/:id/result", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const sessionId = Number(c.req.param("id"));

  const result = await getSessionResult(db, sessionId, user.id);
  return c.json(result);
});

// GET /exam-sessions/:id/questions/:orderIndex
sessionRoutes.get("/:id/questions/:orderIndex", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const sessionId = Number(c.req.param("id"));
  const orderIndex = Number(c.req.param("orderIndex"));

  const result = await getSessionQuestion(db, sessionId, orderIndex, user.id);
  return c.json(result);
});

// POST /exam-sessions/:id/questions/:orderIndex/submit
sessionRoutes.post("/:id/questions/:orderIndex/submit", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const sessionId = Number(c.req.param("id"));
  const orderIndex = Number(c.req.param("orderIndex"));
  const body = await c.req.json<{ selected_option_ids: number[] }>();
  if (!Array.isArray(body.selected_option_ids)) {
    throw new AppError(400, "selected_option_ids must be an array");
  }

  const result = await submitSessionAnswer(db, sessionId, orderIndex, user.id, body.selected_option_ids);
  return c.json(result);
});

// POST /exam-sessions/:id/complete
sessionRoutes.post("/:id/complete", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const sessionId = Number(c.req.param("id"));

  const result = await completeSession(db, sessionId, user.id);
  return c.json(result);
});

// POST /exam-sessions/:id/abandon
sessionRoutes.post("/:id/abandon", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const sessionId = Number(c.req.param("id"));

  await abandonSession(db, sessionId, user.id);
  return c.body(null, 204);
});

export default sessionRoutes;
